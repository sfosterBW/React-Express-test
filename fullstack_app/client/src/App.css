.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}

#fruit-form {
  background-color: grey;
  padding: 5vmin 8vmin;
  border-radius: 0.5vmin;
}

.input-wrapper {
  display: flex;
  margin-bottom: 2vmin;
}

.input-wrapper label {
  display: inline-block;
  width: 20vmin;
  margin-right: 2vmin;
  font-weight: bold;
  text-align: right;
}

.input-wrapper input[type=text] {
  height: 4vmin;
  font-size: 4vmin;
}

.input-wrapper input[type=radio] {
  margin-left: 1vmin;
  margin-right: 4vmin;
}

#fruit-form button {
  display: flex;
  margin: 0 auto;
  padding: 2vmin 4vmin;
  font-size: 2vmin;
  border-radius: 0.5vmin;
}

#fruit-wrapper {

}

#fruit-wrapper table, th, td {
  border: 1px solid black;
  min-width: 20vmin;
  background-color: white;
  color: black;
  margin: 4vmin;
}

#fruit-wrapper table tr td p{
  margin: 2vmin;
}

#fruit-wrapper .checkbox {
  margin-right: 12px;
}
