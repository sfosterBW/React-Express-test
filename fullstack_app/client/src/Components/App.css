.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}

#fruit-form {
  background-color: grey;
  padding: 5vmin 8vmin;
  border-radius: 0.5vmin;
}

#fruit-form .input-wrapper {
  display: flex;
  margin-bottom: 2vmin;
}

#fruit-form .input-wrapper label {
  display: inline-block;
  width: 20vmin;
  margin-right: 2vmin;
  font-weight: bold;
  text-align: right;
}

#fruit-form .input-wrapper input[type=text] {
  height: 4vmin;
  font-size: 3vmin;
}

#fruit-form .input-wrapper input[type=radio] {
  margin-left: 1vmin;
  margin-right: 4vmin;
}

#fruit-form button {
  display: flex;
  margin: 0 auto;
  padding: 2vmin 4vmin;
  font-size: 2vmin;
  border-radius: 0.5vmin;
}

#fruit-table {}

#fruit-table table {
  border-collapse: collapse;
}

#fruit-table thead {
  border-bottom: 1px solid black;
}

#fruit-table table, th, td {
  min-width: 20vmin;
  background-color: white;
  color: black;
  margin: 4vmin;
}

#fruit-table tbody tr {
  border-bottom: 1px solid lightgrey;
}

#fruit-table table tr td p {
  margin: 2vmin;
}

#fruit-table .checkbox {
  margin-right: 12px;
}
