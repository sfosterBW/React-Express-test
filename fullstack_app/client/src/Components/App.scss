//Global variables
$primary-background: #282c34;
$primary-text: white;
$primary-link-text: #09d3ac;
$primary-shadow: rgba(0,0,0,0.2);
$secondary-background: white;
$secondary-text: black;
$secondary-dark-border: black;
$secondary-light-border: lightgrey;

.App {
  text-align: center;

  .App-header {
    background-color: $primary-background;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);
    color: $primary-text;

    form {
      background: grey;
    }
  }
}

.App-link {
  color: $primary-link-text;
}

.fruit-form {
  padding: 5vmin 8vmin;
  border-radius: 0.5vmin;

  button {
    display: flex;
    margin: 0 auto;
    padding: 2vmin 4vmin;
    font-size: 2vmin;
    border-radius: 0.5vmin;
  }
}

.input-wrapper {
  display: flex;
  margin-bottom: 2vmin;

  label {
    display: inline-block;
    width: 20vmin;
    margin-right: 2vmin;
    font-weight: bold;
    text-align: right;
  }

  input[type=text] {
    height: 4vmin;
    font-size: 3vmin;
  }

  input[type=radio] {
    margin-left: 1vmin;
    margin-right: 4vmin;
  }
}

table {
  border-collapse: collapse;
  background-color: $secondary-background;
  color: black;
  margin: 4vmin;

  thead {
    border-bottom: 1px solid $secondary-dark-border;
  }

  th, td {
    min-width: 20vmin;
    margin: 4vmin;
    padding: 2vmin;
  }

  tbody tr {
    border-bottom: 1px solid $secondary-light-border;
  }

  input[type="checkbox"] {
    margin-right: 12px;
  }

}

.modal-wrapper {
  position: absolute;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
  background: $primary-shadow;
  display: flex;
  justify-content: center;
  align-items: center;

  .modal {
    position: relative;
    border-radius: 5px;
    background: $secondary-background;
    color: $secondary-text;
  }

  .closeButton {
    position: absolute;
    top: 1.5vmin;
    right: 0px;
    transform: translateY(-50%);
    cursor: pointer;
  }
}
