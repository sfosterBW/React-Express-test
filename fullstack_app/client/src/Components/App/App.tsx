import React, { FC, useEffect, useCallback } from 'react'
import { Router } from "@reach/router"
import { useSelector, useDispatch } from 'react-redux'
import { RootState } from '../../utils/store'
import fruitService from '../../utils/api'
import { getFruits } from '../../utils/actions'
import Modal from '../Modal/Modal'
import Nav from '../Nav/Nav'

import AddFruit from '../../Pages/AddFruit'
import EditFruit from '../../Pages/EditFruit'
import Home from '../../Pages/Home'
import styles from './App.module.css'

const App: FC = () => {

  const selectFruits = (state: RootState) => state.fruit.data
  const fruits = useSelector(selectFruits)
  const selectModalFruit = (state: RootState) => state.modal.fruit
  const modalFruit = useSelector(selectModalFruit)
  const dispatch = useDispatch()
  //Used to ensure the same values are used on re-render
  const effectDispatch = useCallback(dispatch, [])

  useEffect(() => {
    const initFruit = () => {
      fruitService.fetchFruit()
        .then(res => effectDispatch(getFruits(res)))
    }
    initFruit()
  }, [effectDispatch])

  return (
    <div className="App">
      <Nav />
      <main className={styles.main}>
        <Router className={styles.main}>
          <Home path="/" title="Add Fruit and stuff" />
          <AddFruit path="/add" title="Add Fruit and stuff" />
          <EditFruit fruits={fruits} path="/edit" title="True table"/>
        </Router>
      </main>
      <Modal fruit={modalFruit} />
    </div>
  )
}

export default App
