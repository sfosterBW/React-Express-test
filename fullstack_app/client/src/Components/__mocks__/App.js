const testData = {
  falseCase: {
    _id: 1,
    name: "False Case",
    best: false
  },
  trueCase: {
    _id: 2,
    name: "True Case",
    best: true
  }
}

export default function getFruitList() {
  return testData.falseCase
}
