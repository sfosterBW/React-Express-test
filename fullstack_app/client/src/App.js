import React, { Component } from 'react';
import './App.css';
import axios from 'axios';

class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      fruitList: [],
    };
    this.displayFruitList = this.displayFruitList.bind(this)
    this.addFruitToList = this.addFruitToList.bind(this)
    this.updateFruitInList = this.updateFruitInList.bind(this)
    this.handleChange = this.handleChange.bind(this)
  }


  getFruitList() {
    axios.get("/fruit-api/list")
      .then(res => {
        const fruitList = res.data
        this.setState({ fruitList: fruitList })
      })
      .catch(e => console.log(e))
  }

  //Convert to it's own child component
  displayFruitList(best) {
      return this.state.fruitList
        .filter(i => i.best === best)
        .map(i => <li key={i._id + i.name}><input id={i._id} class="checkbox" name={i._id + i.name} checked={i.best} type="checkbox"onChange={this.handleChange}/><label htmlFor={i.name}>{i.name}</label></li>)
  }

  async addFruitToList() {
    let newFruit = document.getElementById("fruitInput").value
    const res = await axios.post("/fruit-api/new", { new: newFruit })
    this.getFruitList()
    return await res;
  }

  async updateFruitInList(id = 0, best = false) {
    let updateFruit = id
    const res = await axios.put("/fruit-api/update", { id: updateFruit, value: best })
    this.getFruitList()
    return await res;
  }

  handleChange(e) {
    const target = e.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const id = Number(target.id);
    this.updateFruitInList(id, value)
  }

  componentWillMount() {
    this.getFruitList()
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h2>List out your favourite fruit</h2>
          <input id="fruitInput"></input>
          <button onClick={this.addFruitToList}>Add the best new fruit</button>
          <h3>Best fruit</h3>
            <ul>
              {this.displayFruitList(true)}
            </ul>
          <h3>Not best fruit</h3>
            <ul>
              {this.displayFruitList(false)}
            </ul>
        </header>
      </div>
    );
  }
}

export default App;
