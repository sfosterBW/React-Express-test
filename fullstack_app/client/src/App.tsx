import React, { Component } from 'react';
import './App.css';
import axios from 'axios';

interface Props {
  title: string;
}

interface State {
  fruitList: Fruit[];
  addValue: string;
}

interface Fruit {
  _id: number,
  name: string,
  best: boolean;
}

class App extends Component<Props, State> {

  constructor(props: Props) {
    super(props);
    this.state = {
      fruitList: [],
      addValue: ""
    };
    this.displayFruitList = this.displayFruitList.bind(this)
    this.addFruitToList = this.addFruitToList.bind(this)
    this.updateFruitInList = this.updateFruitInList.bind(this)
    this.handleChange = this.handleChange.bind(this)
    this.handleAddChange = this.handleAddChange.bind(this)
  }


  getFruitList() {
    axios.get("/fruit-api/list")
      .then(res => {
        const fruitList: Fruit[] = res.data
        this.setState({ fruitList: fruitList })
      })
      .catch(e => console.log(e))
  }

  //Convert to it's own child component
  displayFruitList(best: boolean) {
    return this.state.fruitList
      .filter((i: Fruit) => i.best === best)
      .map((i: Fruit) =>
        <tr key={i._id.toString() + i.name}>
          <td>
            <label htmlFor={`${i._id}${i.name}`}>{i.name}</label>
          </td>
          <td>
            <p>{i.best.toString()}</p>
          </td>
          <td>
            <input
              id={i._id.toString()}
              className="checkbox"
              name={`${i._id}${i.name}`}
              checked={i.best}
              type="checkbox"
              onChange={this.handleChange}
            />
          </td>
      </tr>
      )
  }

  async addFruitToList() {
    let newFruit = this.state.addValue
    console.log(newFruit)
    let res = await axios.post("/fruit-api/new", { new: newFruit })
    this.getFruitList()
    return res;
  }

  handleAddChange(e: any) {
    const target = e.target
    const value = target.value
    this.setState({ addValue: value })
  }

  async updateFruitInList(id = 0, best = false) {
    let updateFruit = id
    let res = await axios.put("/fruit-api/update", { id: updateFruit, value: best })
    this.getFruitList()
    return res;
  }

  handleChange(e: any) {
    const target = e.target
    const value = target.type === 'checkbox' ? target.checked : target.value
    const id: number = Number(target.id)
    this.updateFruitInList(id, value)
  }

  componentWillMount() {
    this.getFruitList()
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1>Fruit dashboard</h1>
          <div id="fruit-form">
            <h2>List out your favourite fruit</h2>
            <div className="input-wrapper">
              <label htmlFor="fruitName">Add a fruit:</label>
              <input id="fruitName" name="fruitName" type="text" onChange={this.handleAddChange} />
            </div>
            <div className="input-wrapper">
              <label htmlFor="fruitBest">Best:</label>
              True<input type="radio" name="fruitBest" value="true" onChange={this.handleChange} />
              False<input type="radio" name="fruitBest" value="false" onChange={this.handleChange} />
            </div>
            <button onClick={this.addFruitToList}>Add the best new fruit</button>
          </div>
          <div id="fruit-wrapper">
            <table>
              <tr>
                <th>Fruit</th>
                <th>Best?</th>
                <th>Edit</th>
              </tr>
                  {this.displayFruitList(true)}
                  {this.displayFruitList(false)}
            </table>
          </div>
        </header>
      </div>
    );
  }
}

export default App;
