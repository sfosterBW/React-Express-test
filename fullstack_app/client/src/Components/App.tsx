import React, { Component } from 'react';
import './App.css';
import Form from './Form'
import Row from './Row'
import Table from './Table'
import { Fruit } from './interfaces'
import axios from 'axios';

interface Props {
  title: string;
}

interface State {
  fruitList: Fruit[];
  newFruitName: string;
  newFruitBest: boolean;
}

class App extends Component<Props, State> {

  constructor(props: Props) {
    super(props);
    this.state = {
      fruitList: [],
      newFruitName: "",
      newFruitBest: false,
    };
    this.displayFruitList = this.displayFruitList.bind(this)
    this.addFruit = this.addFruit.bind(this)
    this.updateFruit = this.updateFruit.bind(this)
    this.handleEdit = this.handleEdit.bind(this)
    this.handleAdd = this.handleAdd.bind(this)
  }

  getFruitList() {
    axios.get("/fruit-api/list")
      .then(res => {
        const fruitList: Fruit[] = res.data
        this.setState({ fruitList: fruitList })
      })
      .catch(e => console.log(e, "get fruit list"))
  }

  //Convert to it's own child component
  displayFruitList(best: boolean) {
    return this.state.fruitList
      .filter((i: Fruit) => i.best === best)
      .map((i: Fruit) =>
        <Row
          key={`${i._id}${i.name}`}
          fruit={i}
          onChange={this.handleEdit}/>)
  }

  async addFruit() {
    const name = this.state.newFruitName
    const best = this.state.newFruitBest
    const newFruit = { name: name, best: best }
    if (name !== "") {
      try{
        let res = await axios.post("/fruit-api/new", { new: newFruit })
        this.getFruitList()
        this.setState({ newFruitName: "" })
        return res
      } catch(e) {
        console.log(e, "Add fruit")
      }
    } else {
      alert("Put in a name. You didn't put in a name. Why not?")
    }
  }

  handleAdd(e: any) {
    const target = e.target
    const name = target.name
    const value = target.value
    console.log(target, name, value)
    if (name === "newFruitName" && value !== "") {
      this.setState({ newFruitName: value })
    } else if (name === "newFruitBest") {
      const best: boolean = value === "true" ? true : false
      this.setState({ newFruitBest: best })
    }
  }

  async updateFruit(id = 0, best = false) {
    let updateFruit = id
    try{
      let res = await axios.put("/fruit-api/update", { id: updateFruit, value: best })
      this.getFruitList()
    return res;
    } catch(e) {
      console.log(e, "Update fruit")
    }
  }

  handleEdit(e: any) {
    const target = e.target
    const value = target.type === 'checkbox' ? target.checked : target.value
    const id: number = Number(target.id)
    this.updateFruit(id, value)
  }

  componentDidMount() {
    this.getFruitList()
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1>Fruit dashboard</h1>
          <Form
            name={this.state.newFruitName}
            onChange={this.handleAdd}
            onClick={this.addFruit} />
          <Table title="True table" rows={this.displayFruitList(true)} />
          <Table title="False table" rows={this.displayFruitList(false)} />
        </header>
      </div>
    );
  }
}

export default App;
