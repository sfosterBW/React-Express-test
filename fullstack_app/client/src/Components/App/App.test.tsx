import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import App from './App'

const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}))

describe('the app component', () => {
  jest.mock('./../../utils/api')
  const component =  <App />
  const wrapper = mount(component)

  it('renders with the correct structure', () => {
    expect(wrapper).toBeDefined()
    expect(wrapper.find('div.App')).toHaveLength(1)
    expect(wrapper.find('main')).toHaveLength(1)
    expect(wrapper.find('Nav')).toHaveLength(1)
    expect(wrapper.find('Router')).toHaveLength(1)
    expect(wrapper.find('Home')).toHaveLength(1)
    expect(wrapper.find('AddFruit')).toHaveLength(0)
    expect(wrapper.find('EditFruit')).toHaveLength(0)
    expect(wrapper.find('Modal')).toHaveLength(1)
  })

  it('navigates through pages', () => {
    wrapper.find('nav a').at(2).simulate('click')
    expect(wrapper).toBeDefined()
  })

  it('renders the same as last time', () => {
    const tree = renderer
      .create(component)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
