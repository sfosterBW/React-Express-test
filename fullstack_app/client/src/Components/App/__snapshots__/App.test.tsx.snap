// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the app component renders the same as last time 1`] = `
<div
  className="App"
>
  <header>
    <nav
      className="wrapper"
    >
      <div
        className="left"
      >
        <a
          aria-current="page"
          href="/"
          onClick={[Function]}
        >
          <img
            alt="logo"
            className="logo"
            src="https://img.icons8.com/cotton/64/000000/apple--v1.png"
          />
        </a>
        <div
          className="title"
          id="title"
        >
          Fruit Dashboard
        </div>
      </div>
      <div
        className="right"
      >
        <a
          aria-current="page"
          className="link"
          href="/"
          onClick={[Function]}
          rel="nofollower noreferrer"
          title="Home page"
        >
          Home
        </a>
        <a
          className="link"
          href="/add"
          onClick={[Function]}
          rel="nofollower noreferrer"
          title="Add a new fruit to your list"
        >
          Add new fruit
        </a>
        <a
          className="link"
          href="/edit"
          onClick={[Function]}
          rel="nofollower noreferrer"
          title="View and edit your list of fruit"
        >
          Edit fruit
        </a>
      </div>
    </nav>
  </header>
  <main
    className="main"
  >
    <div
      className="main"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <section
        className="wrapper"
      >
        <div
          className="hero large flavourOne"
        >
          <h1
            className="title"
          >
            Add Fruit and stuff
          </h1>
        </div>
      </section>
      <div
        className="card-grid-wrapper"
      >
        <h2>
          What does it do?
        </h2>
        <div
          className="card-grid"
        >
          <div
            className="card"
          >
            <h3>
              Add Fruit to the dashboard
            </h3>
            <p>
              You can add any Fruit to the dashboard very easily.
            </p>
          </div>
          <div
            className="card"
          >
            <h3>
              Edit your Fruit
            </h3>
            <p>
              Once you've added fruit you can update any details and remove it if necessary.
            </p>
          </div>
          <div
            className="card"
          >
            <h3>
              Rank the Fruit
            </h3>
            <p>
              You can pick which Fruit is best and then you know which ones are the best ones.
            </p>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
